externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToLinalg/Misc.cpp

```diff
@@ -0,0 +1,100 @@
+//===------------------------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "torch-mlir-dialects/Conversion/TcpToLinalg/TcpToLinalg.h"
+
+#include "../PassDetail.h"
+#include "PopulatePatterns.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "torch-mlir-dialects/Dialect/Tcp/IR/TcpDialect.h"
+#include "torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.h"
+
+using namespace mlir;
+using namespace mlir::tcp;
+
+namespace {
+
+void getValuesFromIndexArrayAttribute(ArrayAttr attr,
+                                      SmallVector<int64_t> &arrayValues) {
+  for (Attribute val : attr.getValue()) {
+    arrayValues.push_back(val.cast<IntegerAttr>().getValue().getSExtValue());
+  }
+}
+
+class ConvertBroadcastOp : public OpConversionPattern<BroadcastOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+              BroadcastOp op,
+              OpAdaptor adaptor,
+              ConversionPatternRewriter &b) const override {
+    Location loc = op->getLoc();
+    auto resultTensorType = OpConversionPattern::getTypeConverter()
+                  ->convertType(op->getResult(0).getType()).template cast<RankedTensorType>();
+    auto inputTensor = op->getOperands()[0];
+
+    SmallVector<int64_t> axes;
+    getValuesFromIndexArrayAttribute(op.getAxes(), axes);
+
+    auto resultRank = resultTensorType.getRank();
+    SmallVector<Value> resultDimSizes;
+    SmallVector<AffineExpr> exprs;
+    int64_t pos = 0;
+    for (int64_t i = 0; i < resultRank; ++i) {
```

Comment:
Some comments here would be nice.

Reply:
Didn't add explicit comments, but refactored the code to be more understandable.  LMK if this still needs comments.